lib/flutter/flutter_daemon.dart

```diff
@@ -0,0 +1,373 @@
+          _domains[prefix]._handleEvent(event, json['params']);
+        }
+      } else {
+        Completer completer = _completers.remove(json['id']);
+        String methodName = _methodNames.remove(json['id']);
+
+        if (completer == null) {
+          _logger.severe('unmatched request response: ${message}');
+        } else if (json['error'] != null) {
+          completer.completeError(new RequestError(methodName, json['error']));
+        } else {
+          completer.complete(json['result']);
+        }
+      }
+    } catch (e) {
+      _logger.severe('unable to decode message: ${message}, ${e}');
+    }
+  }
+
+  Future<dynamic> _call(String method, [Map args]) {
+    String id = '${++_id}';
+    _completers[id] = new Completer();
+    _methodNames[id] = method;
+    Map m = {'id': id, 'method': method};
+    if (args != null) m['params'] = args;
+    String message = _jsonEncoder.encode(m);
+    if (_willSend != null) _willSend(method);
+    _onSend.add(message);
+    _writeMessage(message);
+    return _completers[id].future;
+  }
+
+  static dynamic _toEncodable(obj) => obj is Jsonable ? obj.toMap() : obj;
+}
+
+abstract class Domain {
+  final FlutterDaemon server;
+  final String name;
+
+  Map<String, StreamController> _controllers = {};
+  Map<String, Stream> _streams = {};
+
+  Domain(this.server, this.name) {
+    server._domains[name] = this;
+  }
+
+  Future<dynamic> _call(String method, [Map args]) => server._call(method, args);
+
+  Stream<dynamic> _listen(String name, Function cvt) {
+    if (_streams[name] == null) {
+      _controllers[name] = new StreamController.broadcast();
+      _streams[name] = _controllers[name].stream.map(cvt);
+    }
+
+    return _streams[name];
+  }
+
+  void _handleEvent(String name, dynamic event) {
+    if (_controllers[name] != null) {
+      _controllers[name].add(event);
+    }
+  }
+
+  String toString() => 'Domain ${name}';
+}
+
+abstract class Jsonable {
+  Map toMap();
+}
+
+class RequestError {
+  final String methodName;
+  final dynamic error;
+
+  RequestError(this.methodName, this.error);
+
+  String toString() => '[RequestError ${error}]';
+}
+
+Map _mapify(Map m) {
+  Map copy = {};
+
+  for (var key in m.keys) {
+    var value = m[key];
+    if (value != null) copy[key] = value;
+  }
+
+  return copy;
+}
+
+class DaemonDomain extends Domain {
+  DaemonDomain(FlutterDaemon server) : super(server, 'daemon');
+
+  Stream<LogMessage> get onLogMessage {
+    return _listen('daemon.logMessage', LogMessage.parse);
+  }
+
+  Future<String> version() => _call('daemon.version');
+
+  Future shutdown() => _call('daemon.shutdown');
+}
+
+class AppDomain extends Domain {
+  AppDomain(FlutterDaemon server) : super(server, 'app');
+
+  // TODO: result
+  // TODO: We need the stdout and stderr from launching the process.
+  Future<dynamic> start(
+    String projectDirectory, {
+    String target,
+    bool checked,
+    String route
+  }) {
+    return _call('app.start', _mapify({
```

Comment:
I can rename this method, but what `_mapify` does is return a copy of the map with all keys stripped out that had associated `null` values. So a call site can create a map w/ all possible keys, and `_mapify` will strip out any keys that didn't have associated values.